service/s3/s3manager/buffered_read_seeker.go

```diff
@@ -9,15 +9,20 @@ import (
 // BufferedReadSeeker is buffered io.ReadSeeker
 type BufferedReadSeeker struct {
 	r                 io.ReadSeeker
-	buffer            []byte
+	buffer            *[]byte
 	readIdx, writeIdx int
 }
 
 // NewBufferedReadSeeker returns a new BufferedReadSeeker
 // if len(b) == 0 then the buffer will be initialized to 64 KiB.
 func NewBufferedReadSeeker(r io.ReadSeeker, b []byte) *BufferedReadSeeker {
-	if len(b) == 0 {
-		b = make([]byte, 64*1024)
+	return newBufferedReadSeeker(r, &b)
+}
+
+func newBufferedReadSeeker(r io.ReadSeeker, b *[]byte) *BufferedReadSeeker {
+	if b == nil || len(*b) == 0 {
+		bs := make([]byte, 64*1024)
```

Comment:
Could we have used `bytes.Buffer` here?

Reply:
No, using `bytes.Buffer` will try to grow the buffer which is not the behavior we want in this case.